File: python_src/kaudio_app/nodes/effect/volume_node.py
from kaudio_app.nodes.abstract.base_node import BaseNode
import kaudio


class VolumeNode(BaseNode):
    NODE_NAME = "Volume"

    def __init__(self):
        self.gain = 0.0

        super().__init__()

        self.add_float_input("gain", "gain (dB)", 0.0, (-10.0, 10.0))

    def get_new_node(self, stereo: bool) -> kaudio.BaseNode:
        node = kaudio.VolumeNode(stereo)
        node.gain = self.gain
        return node

    def set_property(self, name, value):
        if name == "gain":
            self.gain = value
            self.node.gain = value
        else:
            super().set_property(name, value)

File: src/nativeMain/kotlin/_kaudio/nodes/effect/VolumeNode.kt
package _kaudio.nodes.effect

import _kaudio.FRAME_SIZE
import _kaudio.nodes.abstract.DualNode
import _kaudio.nodes.abstract.PyType_DualNode
import _kaudio.utils.dbToAmp
import kotlinx.cinterop.*
import python.KtPyObject
import pywrapper.PyObjectT
import pywrapper.builders.makePyType
import pywrapper.ext.arg
import pywrapper.ext.parseKw

class VolumeNode(stereo: Boolean) : DualNode(stereo) {
    private val gain by attribute("gain", 0f)

    override fun processMono() {
        val amp = dbToAmp(gain)
        val inp = input
        val out = output
        for (i in 0 until FRAME_SIZE) {
            out[i] = inp[i] * amp
        }
    }

    override fun processStereo() {
        val amp = dbToAmp(gain)

        val outL = outputLeft
        val outR = outputRight
        val inL = inputLeft
        val inR = inputRight

        for (i in 0 until FRAME_SIZE) {
            outL[i] = inL[i] * amp
            outR[i] = inR[i] * amp
        }
    }
}

private val initVolumeNode = staticCFunction { self: PyObjectT, args: PyObjectT, kwargs: PyObjectT ->
    memScoped {
        val selfObj: CPointer<KtPyObject> = self?.reinterpret() ?: return@memScoped -1

        val parsed = args.parseKw("__init__", kwargs, "stereo")
        val instance = VolumeNode(parsed.arg("stereo"))
        val ref = StableRef.create(instance)
        selfObj.pointed.ktObject = ref.asCPointer()
        return@memScoped 0
    }
}

val PyType_VolumeNode = makePyType<VolumeNode>(
    ktp_base = PyType_DualNode.ptr,
    ktp_init = initVolumeNode
)
